klaus/contrib/auto_klaus.py

```diff
@@ -0,0 +1,103 @@
+"""
+Alternative take on the "automatically discovered repositories" concept
+that requires no threads, polling or inotify. Instead the filesystem is
+consulted whenever a repository name is looked up.
+
+Since os.path.exists() and os.listdir() are fairly quick filesystem
+operations, performance should be good for small to medium sites.
+FancyRepo() objects are cached.
+
+Repositories are identified by the existence of a
+
+    <reponame>/git-daemon-export-ok
+
+file (for compatibility with gitweb).
+
+Example usage:
+
+    from klaus.contrib.auto_klaus import Klaus, SlashDynamicRepos
+
+    application = Klaus('/srv/git', "My git repositories", False)
+
+    application.wsgi_app = httpauth.AlwaysFailingAuthMiddleware(
+        wsgi_app=dulwich.web.make_wsgi_chain(
+            backend=dulwich.server.DictBackend(SlashDynamicRepos(application.valid_repos)),
+            fallback_app=application.wsgi_app,
+        ),
+    )
+"""
+
+import collections.abc
+import pathlib
+import os
+import os.path
+
+import klaus
+import klaus.repo
+
+class FilesystemRepoDict(collections.abc.Mapping):
+    """
+    Maintain a virtual read-only dictionary whose contents represent
+    the presence of git repositories in the given root directory.
+    """
+
+    def __init__(self, root):
+        self._root = pathlib.Path(root)
+        self._repos = {}
+
+    def __getitem__(self, name):
+        if not name or name[0] == '.' or '/' in name or '\0' in name:
+            raise KeyError(name)
+
+        repos = self._repos
+        path = self._root / name
+        if not os.path.exists(path / 'git-daemon-export-ok'):
+            repos.pop(name, None)
+            raise KeyError(name)
+
+        try:
+            return repos[name]
+        except KeyError:
+            pass
+
+        repo = klaus.repo.FancyRepo(str(path))
+        repos[name] = repo
+        return repo
+
+    def __iter__(self):
+        root = self._root
+        return (
+            repo for repo in os.listdir(root)
+                if os.path.exists(root / repo / 'git-daemon-export-ok')
```

Comment:
Could we save a few `exists()` checks here by checking `self._repos` for a positive match first?

Reply:
That's what I had at first, in fact. But then the admin cannot retract access from a repository. I suppose we could make that optional.

Reply:
Right but if we keep the safety check + `.pop()` in `__getitem__` that's an acceptable compromise IMO.

Reply:
I don't follow. The “safety check + `.pop()`” is what we have now. The alternative is to remove the safety check and just rely on the presence of `self._repos[repo]`. I don't see the compromise you're trying to get at.

Reply:
Added the `detect_removal` option as a compromise between performance and correctness.